cmd/exec/exec.go

```diff
@@ -77,22 +53,136 @@ func getCommandFromName(config config.OneBuildConfiguration, cmd string) string
 	return ""
 }
 
-func createMapOfCommandsToExecute(oneBuildConfiguration config.OneBuildConfiguration, commands ...string) (map[string]string, error) {
-	var cmdMap map[string]string
-	cmdMap = make(map[string]string)
+func bashCommand(s *sh.Session, command string) *sh.Session {
+	return s.Command("bash", "-c", command)
+}
+
+func executeAndStopIfFailed(command *CommandContext) {
+	err := command.commandSession.Run()
+	if err != nil {
+		exitCode := (err.Error())[12:]
+		utils.PrintlnDashedErr("Execution failed during phase \"" + command.name + "\" - Execution of the script \"" + command.command + "\" returned non-zero exit code : " + exitCode)
+		utils.ExitWithCode(exitCode)
+	}
+}
+
+func buildExecutionPlan(onebuildConfig config.OneBuildConfiguration, commands ...string) OneBuildExecutionPlan {
+
+	before := onebuildConfig.Before
+	var executionPlan OneBuildExecutionPlan
+	if before != "" {
+		executionPlan.Init = &CommandContext{"before", before, bashCommand(sh.NewSession(), before)}
+	}
 
 	for _, name := range commands {
-		value := getCommandFromName(oneBuildConfiguration, name)
-		if value == "" {
-			utils.Println("No command '" + name + "' found in config file\n")
-			config.PrintConfiguration(oneBuildConfiguration)
-			return nil, errors.New("")
+		executionCommand := getCommandFromName(onebuildConfig, name)
+		if executionCommand == "" {
+			utils.PrintlnErr("Error building exectuion plan. Command \"" + name + "\" not found.")
+			config.PrintConfiguration(onebuildConfig)
+			utils.Exit(1)
 		}
-		cmdMap[name] = value
+		executionPlan.Commands = append(executionPlan.Commands, &CommandContext{name, executionCommand, bashCommand(sh.NewSession(), executionCommand)})
+	}
+
+	after := onebuildConfig.After
+	if after != "" {
+		executionPlan.Destroy = &CommandContext{"after", after, bashCommand(sh.NewSession(), after)}
 	}
-	return cmdMap, nil
+
+	return executionPlan
+}
+
+// OneBuildExecutionPlan holds all information for the execution strategy
+type OneBuildExecutionPlan struct {
+	Init     *CommandContext
+	Commands []*CommandContext
+	Destroy  *CommandContext
 }
 
-func executeAndStopIfFailed(command string) error {
-	return sh.NewSession().Command("bash", "-c", command).Run()
+// CommandContext holds all meta-data and required information for execution of a command
+type CommandContext struct {
+	name           string
+	command        string
+	commandSession *sh.Session
+}
+
+func (executionPlan *OneBuildExecutionPlan) hasInit() bool {
+	if executionPlan.Init != nil {
+		return true
+	}
+	return false
+}
+
+func (executionPlan *OneBuildExecutionPlan) hasDestroy() bool {
+	if executionPlan.Destroy != nil {
+		return true
+	}
+	return false
+}
+
+func (executionPlan *OneBuildExecutionPlan) hasCommands() bool {
+	if len(executionPlan.Commands) > 0 {
+		return true
+	}
+	return false
+}
+
+func (executionPlan *OneBuildExecutionPlan) print() {
```

Comment:
Can we use `Utils` to print statements not `fmt` everywhere.

The Idea behind using utils is - we can introduce/change any printing library without changing print statements everywhere.

WDYT?